### PR TITLE
Presentation: Complex hierarchy level instance filters

### PR DESCRIPTION
**Pull request by grigasp**
Part of https://github.com/iTwin/itwinjs-backlog/issues/556.
Related to https://github.com/iTwin/itwinjs-core/pull/4864.

Follow-up item: https://github.com/iTwin/itwinjs-core/issues/4863.

### CODE REVIEW

File: iModelJsNodeAddon/presentation/UiStateProvider.cpp [hunk 1]
```diff
@@ -135,10 +135,13 @@ ECPresentationResult IModelJsECPresentationUiStateProvider::UpdateHierarchyState
 
         if (item.isMember(PRESENTATION_JSON_ATTRIBUTE_StateChanges_InstanceFilters) && !item[PRESENTATION_JSON_ATTRIBUTE_StateChanges_InstanceFilters].isArray())
             return ECPresentationResult(ECPresentationStatus::InvalidArgument, "Expected `" PRESENTATION_JSON_ATTRIBUTE_StateChanges_InstanceFilters "` member in `stateChanges` items to be an array of strings");
-        bvector<Utf8String> instanceFilters;
-        item[PRESENTATION_JSON_ATTRIBUTE_StateChanges_InstanceFilters].ForEachArrayMember([&](BeJsConst::ArrayIndex, BeJsConst instanceFilterJson)
+
+        bvector<std::shared_ptr<InstanceFilterDefinition const>> instanceFilters;
+        item[PRESENTATION_JSON_ATTRIBUTE_StateChanges_InstanceFilters].ForEachArrayMember([&](BeJsConst::ArrayIndex filterIndex, BeJsConst instanceFilterJson)
```
Comment (saskliutas) [inline iModelJsNodeAddon/presentation/UiStateProvider.cpp:140]: Does not look like `filterIndex` variable is used.

File: iModelJsNodeAddon/presentation/ECPresentationSerializer.cpp [hunk 2]
```diff
@@ -1322,3 +1322,158 @@ RulesetVariables IModelJsECPresentationSerializer::GetRulesetVariablesFromJson(B
...
+        if (instanceDef.hasMember(PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_Alias) && instanceDef[PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_Alias].isString())
+            {
+            Utf8String alias = instanceDef[PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_Alias].ToUtf8CP();
+            pathToProperty.back().GetTargetClass().SetAlias(alias);
+            hasAlias = true;
+            }
+
+        if (instanceDef.hasMember(PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_RelationshipAlias) && instanceDef[PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_RelationshipAlias].isString())
+            {
+            Utf8String alias = instanceDef[PRESENTATION_JSON_ATTRIBUTE_InstanceFilterDefinition_RelatedInstances_RelationshipAlias].ToUtf8CP();
```
Comment (saskliutas) [inline iModelJsNodeAddon/presentation/ECPresentationSerializer.cpp:1410]: I think this `if` missing `hasAlias = true` statement. I seed that it was missing it before but I think `hasAlias` flag is here to determine if atleast one of the aliases is present.

File: iModelCore/ECPresentation/Source/Hierarchies/DataSourceInfo.h [hunk 3]
```diff
@@ -111,24 +111,23 @@ struct DataSourceIdentifier
...
         {}
-    bool operator<(DataSourceIdentifier const& other) const
-        {
-        NUMERIC_LESS_COMPARE(m_id, other.m_id);
-        NUMERIC_LESS_COMPARE(m_hierarchyLevelId, other.m_hierarchyLevelId);
-        VECTOR_LESS_COMPARE(m_index, other.m_index);
-        STR_LESS_COMPARE(m_instanceFilter.c_str(), other.m_instanceFilter.c_str());
-        return false;
-        }
+    // bool operator<(DataSourceIdentifier const& other) const
```
Comment (saskliutas) [inline iModelCore/ECPresentation/Source/Hierarchies/DataSourceInfo.h:124]: I guess this is not needed and can be removed?

File: iModelJsNodeAddon/presentation/UiStateProvider.cpp (hunk 1 above)
Comment (grigasp) [reply iModelJsNodeAddon/presentation/UiStateProvider.cpp]: Yes, cleaned up

File: iModelCore/ECPresentation/Source/Hierarchies/DataSourceInfo.h (hunk 3 above)
Comment (grigasp) [reply iModelCore/ECPresentation/Source/Hierarchies/DataSourceInfo.h]: Yes... Fixed.

File: iModelJsNodeAddon/presentation/ECPresentationSerializer.cpp (hunk 2 above)
Comment (grigasp) [reply iModelJsNodeAddon/presentation/ECPresentationSerializer.cpp:1409]: Agree, made a change to set `hasAlias = true` when there is a relationship alias

**Review by saskliutas — APPROVED** [review]

**Pull request merged**